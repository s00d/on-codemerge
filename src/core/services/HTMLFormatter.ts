import type { FormattingOptions } from '../types';

export class HTMLFormatter {
  private defaultOptions: FormattingOptions = {
    indentSize: 2,
    maxLineLength: 80,
  };

  constructor(private options: Partial<FormattingOptions> = {}) {
    this.options = { ...this.defaultOptions, ...options };
  }

  public format(html: string): string {
    let formatted = '';
    let indent = 0;
    const { indentSize } = { ...this.defaultOptions, ...this.options };

    // Split by tags but preserve attributes with > or < symbols
    const parts = html.split(/(<[^>]+>)/g).filter(Boolean);

    parts.forEach((part) => {
      if (!part.trim()) return;

      // Handle closing tags
      if (part.match(/^<\//)) {
        indent = Math.max(0, indent - 1);
      }

      // Add formatted element with proper indentation
      if (part.startsWith('<')) {
        formatted += '\n' + ' '.repeat(indent * indentSize) + part;
      } else {
        formatted += part;
      }

      // Increase indent for opening tags that aren't self-closing or void elements
      if (part.match(/^<[^/][^>]*[^/]>$/) && !this.isVoidElement(part)) {
        indent = Math.max(0, indent + 1);
      }
    });

    return formatted.trim();
  }

  private isVoidElement(tag: string): boolean {
    const match = tag.match(/<([a-zA-Z0-9-]+)/);
    if (!match) return false;

    const voidElements = [
      'area',
      'base',
      'br',
      'col',
      'embed',
      'hr',
      'img',
      'input',
      'link',
      'meta',
      'param',
      'source',
      'track',
      'wbr',
    ];
    return voidElements.includes(match[1].toLowerCase());
  }
}
