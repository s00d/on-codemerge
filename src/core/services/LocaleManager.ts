export class LocaleManager {
  private currentLocale: string = 'en'; // Текущая локаль по умолчанию
  private locales: Record<string, Record<string, string>> = {}; // Словарь локалей
  private fallbackLocale: string = 'en'; // Локаль по умолчанию, если перевод не найден

  constructor(defaultLocale: string = 'en') {
    this.currentLocale = defaultLocale;
    this.fallbackLocale = defaultLocale;
    this.loadLocale(defaultLocale); // Загружаем локаль по умолчанию
  }

  /**
   * Асинхронно загружает локаль из файла через динамический импорт
   * @param locale - Код локали (например, 'en', 'ru')
   */
  public async loadLocale(locale: string): Promise<void> {
    try {
      const translations = await import(`./locales/${locale}.json`);
      this.locales[locale] = translations.default;
    } catch (error) {
      console.error(`Error loading locale "${locale}":`, error);
      if (!this.locales[this.fallbackLocale]) {
        await this.loadLocale(this.fallbackLocale);
      }
    }
  }

  /**
   * Устанавливает текущую локаль и загружает её, если она ещё не загружена
   * @param locale - Код локали (например, 'en', 'ru')
   */
  public async setLocale(locale: string): Promise<void> {
    if (!this.locales[locale]) {
      await this.loadLocale(locale);
    }
    if (this.locales[locale]) {
      this.currentLocale = locale;
    } else {
      console.warn(`Locale "${locale}" is not loaded.`);
    }
  }

  /**
   * Возвращает текущую локаль
   */
  public getCurrentLocale(): string {
    return this.currentLocale;
  }

  /**
   * Получает перевод по ключу
   * @param key - Ключ перевода
   * @param params - Параметры для подстановки в перевод (опционально)
   */
  public translate(key: string, params: Record<string, string> = {}): string {
    const translation =
      this.locales[this.currentLocale]?.[key] || this.locales[this.fallbackLocale]?.[key] || key;
    return Object.keys(params).reduce((acc, paramKey) => {
      return acc.replace(new RegExp(`{{${paramKey}}}`, 'g'), params[paramKey]);
    }, translation);
  }

  /**
   * Возвращает все загруженные локали
   */
  public getLoadedLocales(): string[] {
    return Object.keys(this.locales);
  }
}
