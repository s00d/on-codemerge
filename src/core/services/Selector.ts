// TableSelection.ts (дополнение)
export class Selector {
  private container: HTMLElement;

  constructor(container: HTMLElement) {
    this.container = container;
  }

  private lastSelection: Range | null = null;
  private lastTable: HTMLTableElement | null = null;
  private selectedCells: HTMLTableCellElement[] = [];

  public saveSelection(): void {
    const selection = window.getSelection();

    if (selection && selection.rangeCount > 0) {
      this.lastSelection = selection.getRangeAt(0).cloneRange();
    } else {
      const range = document.createRange();
      const container = this.container;

      if (container && container.childNodes.length > 0) {
        range.selectNodeContents(container);
        range.collapse(false);
      } else {
        range.setStart(container, 0);
        range.collapse(true);
      }

      selection?.removeAllRanges();
      selection?.addRange(range);
      this.lastSelection = range.cloneRange();
    }
  }

  public saveTable(table: HTMLTableElement): void {
    this.lastTable = table;
  }

  public selectCell(cell: HTMLTableCellElement): void {
    this.selectedCells.push(cell);
    cell.classList.add('selected');
  }

  public clearSelectedCells(): void {
    this.selectedCells.forEach((cell) => cell.classList.remove('selected'));
    this.selectedCells = [];
  }

  public getSelectedCells(): HTMLTableCellElement[] {
    return this.selectedCells;
  }

  public restoreSelection(container: HTMLElement): Range | null {
    if (!this.lastSelection) return null;

    // Проверяем, что начальный и конечный узлы диапазона находятся внутри контейнера
    const startContainer = this.lastSelection.startContainer;
    const endContainer = this.lastSelection.endContainer;

    if (!container.contains(startContainer) || !container.contains(endContainer)) {
      this.clearSelection(); // Очищаем сохранённое выделение, если оно вне контейнера
      return null;
    }

    // Восстанавливаем выделение
    const selection = window.getSelection();
    if (selection) {
      selection.removeAllRanges();
      selection.addRange(this.lastSelection);
    }

    return this.lastSelection;
  }

  public restoreTable(): HTMLTableElement | null {
    return this.lastTable;
  }

  public clearSelection(): void {
    this.lastSelection = null;
  }

  public clearTable(): void {
    this.lastTable = null;
  }
}
