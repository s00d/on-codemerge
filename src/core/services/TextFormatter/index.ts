import { DomUtils } from './DomUtils';
import { StyleManager } from './StyleManager';

export class TextFormatter {
  private domUtils: DomUtils;
  private readonly styleManager: StyleManager;

  constructor(private container: HTMLElement) {
    this.domUtils = new DomUtils(container);
    this.styleManager = new StyleManager();
  }

  toggleStyle(styleCommand: string): void {
    const selection = window.getSelection();
    if (!selection) return;

    const nodesToStyle = this.domUtils.getSelectedRoot(selection);
    if (!nodesToStyle) return;

    nodesToStyle.forEach((node) => {
      // Обернуть текстовый узел в <span>, если это необходимо
      if (node.nodeType === Node.TEXT_NODE) {
        node = this.wrapTextNodeInSpan(node);
      }

      // Применить или удалить стиль
      const element = node as HTMLElement;
      const isFormatted = this.styleManager.has(element, styleCommand);

      if (isFormatted) {
        this.domUtils.removeStyleFromNode(
          node,
          styleCommand,
          this.styleManager.remove.bind(this.styleManager)
        );
      } else {
        this.domUtils.applyStyleToNode(
          node,
          styleCommand,
          this.styleManager.set.bind(this.styleManager)
        );
      }
    });
  }

  setColor(color: string): void {
    this.applyStyleToSelectedNodes((element) => {
      element.style.color = color;
      element.classList.add('format');
    });
  }

  setBackgroundColor(color: string): void {
    this.applyStyleToSelectedNodes((element) => {
      element.style.backgroundColor = color;
      element.classList.add('format');
    });
  }

  setFont(fontFamily: string, fontSize: string): void {
    this.applyStyleToSelectedNodes((element) => {
      element.style.fontFamily = fontFamily;
      element.style.fontSize = fontSize;
      element.classList.add('format');
    });
  }

  clearFont(): void {
    this.applyStyleToSelectedNodes((element) => {
      element.style.removeProperty('font-family');
      element.style.removeProperty('font-size');
      element.classList.add('format');
    });
  }

  private applyStyleToSelectedNodes(styleApplier: (element: HTMLElement) => void): void {
    const selection = window.getSelection();
    if (!selection) return;

    const nodesToStyle = this.domUtils.getSelectedRoot(selection);
    if (!nodesToStyle) return;

    nodesToStyle.forEach((node) => {
      if (node.nodeType === Node.ELEMENT_NODE && node !== this.container) {
        styleApplier(node as HTMLElement);
      }
    });
  }

  private wrapTextNodeInSpan(node: Node): HTMLElement {
    const span = document.createElement('p');
    node.parentNode?.replaceChild(span, node);
    span.appendChild(node);
    return span;
  }

  hasClass(styleCommand: string): boolean {
    const selection = window.getSelection();
    if (!selection || selection.rangeCount === 0) return false;

    const nodesToCheck = this.domUtils.getSelectedRoot(selection, true);
    if (!nodesToCheck) return false;

    // Проверяем, применен ли стиль к любому из выделенных узлов
    return nodesToCheck.some((node) => {
      if (node.nodeType === Node.ELEMENT_NODE) {
        const element = node as HTMLElement;
        return this.styleManager.has(element, styleCommand);
      }
      return false;
    });
  }

  getStyle(name: keyof CSSStyleDeclaration): string | null | undefined {
    const selection = window.getSelection();
    if (!selection || selection.rangeCount === 0) return null;

    const nodesToCheck = this.domUtils.getSelectedRoot(selection, true);
    if (!nodesToCheck) return null;

    // Проверяем, применен ли стиль к любому из выделенных узлов
    for (const nodesToCheckItem of nodesToCheck) {
      if (nodesToCheckItem.nodeType === Node.ELEMENT_NODE) {
        const element = nodesToCheckItem as HTMLElement;
        return (element.style[name] ?? null) as string | null;
      }
      return null;
    }
  }
}
