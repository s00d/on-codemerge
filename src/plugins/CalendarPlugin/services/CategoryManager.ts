import type { Category, Tag } from '../types';

export class CategoryManager {
  private categoriesKey = 'html-editor-calendar-categories';
  private tagsKey = 'html-editor-calendar-tags';

  // Категории
  public getCategories(): Category[] {
    const stored = localStorage.getItem(this.categoriesKey);
    return stored ? JSON.parse(stored) : this.getDefaultCategories();
  }

  public getCategory(id: string): Category | null {
    const categories = this.getCategories();
    return categories.find((cat) => cat.id === id) || null;
  }

  public createCategory(name: string, color: string): Category {
    const categories = this.getCategories();
    const newCategory: Category = {
      id: crypto.randomUUID(),
      name,
      color,
      createdAt: Date.now(),
    };

    categories.push(newCategory);
    localStorage.setItem(this.categoriesKey, JSON.stringify(categories));
    return newCategory;
  }

  public updateCategory(id: string, data: Partial<Category>): Category {
    const categories = this.getCategories();
    const index = categories.findIndex((cat) => cat.id === id);

    if (index === -1) {
      throw new Error('Category not found');
    }

    const updated: Category = {
      ...categories[index],
      ...data,
    };

    categories[index] = updated;
    localStorage.setItem(this.categoriesKey, JSON.stringify(categories));
    return updated;
  }

  public deleteCategory(id: string): void {
    const categories = this.getCategories().filter((cat) => cat.id !== id);
    localStorage.setItem(this.categoriesKey, JSON.stringify(categories));
  }

  // Теги
  public getTags(): Tag[] {
    const stored = localStorage.getItem(this.tagsKey);
    return stored ? JSON.parse(stored) : this.getDefaultTags();
  }

  public getTag(id: string): Tag | null {
    const tags = this.getTags();
    return tags.find((tag) => tag.id === id) || null;
  }

  public createTag(name: string, color: string): Tag {
    const tags = this.getTags();
    const newTag: Tag = {
      id: crypto.randomUUID(),
      name,
      color,
      createdAt: Date.now(),
    };

    tags.push(newTag);
    localStorage.setItem(this.tagsKey, JSON.stringify(tags));
    return newTag;
  }

  public updateTag(id: string, data: Partial<Tag>): Tag {
    const tags = this.getTags();
    const index = tags.findIndex((tag) => tag.id === id);

    if (index === -1) {
      throw new Error('Tag not found');
    }

    const updated: Tag = {
      ...tags[index],
      ...data,
    };

    tags[index] = updated;
    localStorage.setItem(this.tagsKey, JSON.stringify(tags));
    return updated;
  }

  public deleteTag(id: string): void {
    const tags = this.getTags().filter((tag) => tag.id !== id);
    localStorage.setItem(this.tagsKey, JSON.stringify(tags));
  }

  // Утилиты
  public getCategoryByName(name: string): Category | null {
    const categories = this.getCategories();
    return categories.find((cat) => cat.name.toLowerCase() === name.toLowerCase()) || null;
  }

  public getTagByName(name: string): Tag | null {
    const tags = this.getTags();
    return tags.find((tag) => tag.name.toLowerCase() === name.toLowerCase()) || null;
  }

  public getEventsByCategory(categoryId: string, events: any[]): any[] {
    return events.filter((event) => event.category === categoryId);
  }

  public getEventsByTag(tagName: string, events: any[]): any[] {
    return events.filter((event) => event.tags?.includes(tagName));
  }

  // Статистика
  public getCategoryStats(events: any[]): Record<string, number> {
    const stats: Record<string, number> = {};
    const categories = this.getCategories();

    categories.forEach((category) => {
      stats[category.id] = this.getEventsByCategory(category.id, events).length;
    });

    return stats;
  }

  public getTagStats(events: any[]): Record<string, number> {
    const stats: Record<string, number> = {};
    const tags = this.getTags();

    tags.forEach((tag) => {
      stats[tag.id] = this.getEventsByTag(tag.name, events).length;
    });

    return stats;
  }

  // Предустановленные категории и теги
  private getDefaultCategories(): Category[] {
    return [
      { id: 'work', name: 'Work', color: '#3b82f6', createdAt: Date.now() },
      { id: 'personal', name: 'Personal', color: '#10b981', createdAt: Date.now() },
      { id: 'meeting', name: 'Meeting', color: '#f59e0b', createdAt: Date.now() },
      { id: 'travel', name: 'Travel', color: '#8b5cf6', createdAt: Date.now() },
      { id: 'health', name: 'Health', color: '#ef4444', createdAt: Date.now() },
    ];
  }

  private getDefaultTags(): Tag[] {
    return [
      { id: 'urgent', name: 'Urgent', color: '#ef4444', createdAt: Date.now() },
      { id: 'important', name: 'Important', color: '#f59e0b', createdAt: Date.now() },
      { id: 'follow-up', name: 'Follow-up', color: '#3b82f6', createdAt: Date.now() },
      { id: 'completed', name: 'Completed', color: '#10b981', createdAt: Date.now() },
      { id: 'cancelled', name: 'Cancelled', color: '#6b7280', createdAt: Date.now() },
    ];
  }
}
