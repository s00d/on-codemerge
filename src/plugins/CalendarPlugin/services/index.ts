export { CalendarManager } from './CalendarManager';
export { CategoryManager } from './CategoryManager';
export { ReminderService } from './ReminderService';
