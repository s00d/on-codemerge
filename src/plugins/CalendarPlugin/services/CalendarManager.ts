import type {
  Calendar,
  CalendarEvent,
  CreateCalendarData,
  CreateEventData,
  UpdateCalendarData,
  UpdateEventData,
} from '../types';
import { CategoryManager } from './CategoryManager';
import { ReminderService } from './ReminderService';

export class CalendarManager {
  private calendarsKey = 'html-editor-calendars';
  private eventsKey = 'html-editor-calendar-events';
  private categoryManager: CategoryManager;
  private reminderService: ReminderService;

  constructor() {
    this.categoryManager = new CategoryManager();
    this.reminderService = new ReminderService(this);
  }

  public getCalendars(): Calendar[] {
    const stored = localStorage.getItem(this.calendarsKey);
    return stored ? JSON.parse(stored) : [];
  }

  public getCalendar(id: string): Calendar | null {
    const calendars = this.getCalendars();
    return calendars.find((cal) => cal.id === id) || null;
  }

  public createCalendar(data: CreateCalendarData): Calendar {
    const calendars = this.getCalendars();
    const newCalendar: Calendar = {
      id: crypto.randomUUID(),
      title: data.title,
      description: data.description,
      events: data.events ? data.events.map((event) => this.createEvent(event)) : [],
      createdAt: Date.now(),
      updatedAt: Date.now(),
    };

    calendars.push(newCalendar);
    localStorage.setItem(this.calendarsKey, JSON.stringify(calendars));
    return newCalendar;
  }

  public updateCalendar(id: string, data: UpdateCalendarData): Calendar {
    const calendars = this.getCalendars();
    const index = calendars.findIndex((cal) => cal.id === id);

    if (index === -1) {
      throw new Error('Calendar not found');
    }

    const { events, ...otherData } = data;
    const updated: Calendar = {
      ...calendars[index],
      ...otherData,
      events: events ? events.map((event) => this.createEvent(event, id)) : calendars[index].events,
      updatedAt: Date.now(),
    };

    calendars[index] = updated;
    localStorage.setItem(this.calendarsKey, JSON.stringify(calendars));
    return updated;
  }

  public deleteCalendar(id: string): void {
    const calendars = this.getCalendars().filter((cal) => cal.id !== id);
    localStorage.setItem(this.calendarsKey, JSON.stringify(calendars));

    // Удаляем все напоминания календаря
    this.reminderService.deleteEventReminders(id);
  }

  public getEvents(calendarId: string): CalendarEvent[] {
    const allEvents = this.getAllEvents();
    return allEvents.filter((event) => event.calendarId === calendarId);
  }

  public getEvent(id: string): CalendarEvent | null {
    const allEvents = this.getAllEvents();
    return allEvents.find((event) => event.id === id) || null;
  }

  public createEvent(data: CreateEventData, calendarId?: string): CalendarEvent {
    const allEvents = this.getAllEvents();
    const newEvent: CalendarEvent & { calendarId: string } = {
      id: crypto.randomUUID(),
      title: data.title,
      description: data.description,
      date: data.date,
      time: data.time,
      duration: data.duration,
      location: data.location,
      color: data.color,
      isAllDay: data.isAllDay,
      priority: data.priority || 'medium',
      category: data.category,
      tags: data.tags || [],
      attendees: data.attendees || [],
      reminder: data.reminder,
      recurring: data.recurring,
      attachments: data.attachments || [],
      calendarId: calendarId || 'default',
      createdAt: Date.now(),
      updatedAt: Date.now(),
    };

    allEvents.push(newEvent);
    localStorage.setItem(this.eventsKey, JSON.stringify(allEvents));

    // Создаем напоминание, если указано
    if (newEvent.reminder && calendarId) {
      this.reminderService.createReminder(newEvent, calendarId);
    }

    return newEvent;
  }

  public updateEvent(id: string, data: UpdateEventData): CalendarEvent {
    const allEvents = this.getAllEvents();
    const index = allEvents.findIndex((event) => event.id === id);

    if (index === -1) {
      throw new Error('Event not found');
    }

    const oldEvent = allEvents[index];
    const updated: CalendarEvent & { calendarId: string } = {
      ...oldEvent,
      ...data,
      updatedAt: Date.now(),
    };

    allEvents[index] = updated;
    localStorage.setItem(this.eventsKey, JSON.stringify(allEvents));

    // Обновляем напоминание, если изменилось
    if (data.reminder !== undefined && oldEvent.reminder !== data.reminder) {
      this.reminderService.deleteEventReminders(id);
      if (data.reminder && updated.calendarId) {
        this.reminderService.createReminder(updated, updated.calendarId);
      }
    }

    return updated;
  }

  public deleteEvent(id: string): void {
    const allEvents = this.getAllEvents().filter((event) => event.id !== id);
    localStorage.setItem(this.eventsKey, JSON.stringify(allEvents));

    // Удаляем напоминания события
    this.reminderService.deleteEventReminders(id);
  }

  public getEventsByDate(date: string): CalendarEvent[] {
    const allEvents = this.getAllEvents();
    return allEvents.filter((event) => event.date === date);
  }

  public getEventsByDateRange(startDate: string, endDate: string): CalendarEvent[] {
    const allEvents = this.getAllEvents();
    return allEvents.filter((event) => {
      const eventDate = new Date(event.date);
      const start = new Date(startDate);
      const end = new Date(endDate);
      return eventDate >= start && eventDate <= end;
    });
  }

  // Новые методы для работы с категориями и тегами
  public getEventsByCategory(categoryId: string): CalendarEvent[] {
    const allEvents = this.getAllEvents();
    return allEvents.filter((event) => event.category === categoryId);
  }

  public getEventsByTag(tagName: string): CalendarEvent[] {
    const allEvents = this.getAllEvents();
    return allEvents.filter((event) => event.tags?.includes(tagName));
  }

  public getEventsByPriority(priority: 'low' | 'medium' | 'high'): CalendarEvent[] {
    const allEvents = this.getAllEvents();
    return allEvents.filter((event) => event.priority === priority);
  }

  public searchEvents(query: string): CalendarEvent[] {
    const allEvents = this.getAllEvents();
    const lowerQuery = query.toLowerCase();

    return allEvents.filter(
      (event) =>
        event.title.toLowerCase().includes(lowerQuery) ||
        event.description?.toLowerCase().includes(lowerQuery) ||
        event.location?.toLowerCase().includes(lowerQuery) ||
        event.tags?.some((tag) => tag.toLowerCase().includes(lowerQuery))
    );
  }

  // Методы для работы с категориями и тегами
  public getCategoryManager(): CategoryManager {
    return this.categoryManager;
  }

  public getReminderService(): ReminderService {
    return this.reminderService;
  }

  // Генерация HTML с напоминаниями
  public generateCalendarHTMLWithReminders(calendarId: string): string {
    const calendar = this.getCalendar(calendarId);
    if (!calendar) return '';

    const reminderScript = this.reminderService.generateReminderScript(calendarId);

    return `
      ${this.generateCalendarHTML(calendar)}
      ${reminderScript}
    `;
  }

  // Генерация HTML календаря
  public generateCalendarHTML(calendar: Calendar): string {
    const events = this.getEvents(calendar.id);

    // Сортируем события по дате и времени
    const sortedEvents = events.sort((a, b) => {
      const dateA = new Date(`${a.date}T${a.time}`);
      const dateB = new Date(`${b.date}T${b.time}`);
      return dateA.getTime() - dateB.getTime();
    });

    const eventsHtml = sortedEvents.map((event) => this.generateEventHTML(event)).join('');

    return `
      <div class="calendar-widget" data-calendar-id="${calendar.id}">
        <div class="calendar-header">
          <h3 class="calendar-title">${calendar.title}</h3>
        </div>
        <div class="calendar-body">
          <div class="calendar-events">
            ${eventsHtml}
          </div>
        </div>
      </div>
    `;
  }

  // Генерация HTML события
  private generateEventHTML(event: CalendarEvent): string {
    const priorityClass = event.priority || 'medium';
    const categoryColor = event.color || '#3b82f6';
    const categoryName = event.category || 'General';

    // Форматируем дату
    const eventDate = new Date(event.date);
    const formattedDate = eventDate.toLocaleDateString('ru-RU', {
      day: 'numeric',
      month: 'short',
      year: 'numeric',
    });

    const tagsHtml =
      event.tags && event.tags.length > 0
        ? `<div class="event-tags">${event.tags.map((tag) => `<span class="event-tag">${tag}</span>`).join('')}</div>`
        : '';

    const attendeesHtml =
      event.attendees && event.attendees.length > 0
        ? `<span class="event-attendees">👥 ${event.attendees.join(', ')}</span>`
        : '';

    const reminderHtml = event.reminder
      ? `<span class="event-reminder">⏰ ${event.reminder}</span>`
      : '';

    const locationHtml = event.location
      ? `<span class="event-location">📍 ${event.location}</span>`
      : '';

    const durationHtml = event.duration
      ? `<span class="event-duration">⏱️ ${event.duration} min</span>`
      : '';

    const metaInfo = [locationHtml, durationHtml, attendeesHtml, reminderHtml]
      .filter(Boolean)
      .join(' • ');

    return `
      <div class="calendar-event" data-event-id="${event.id}" style="--event-color: ${event.color || '#3b82f6'}">
        <div class="event-header">
          <div class="event-datetime">
            <div class="event-date">${formattedDate}</div>
            <div class="event-time">${event.time}</div>
          </div>
          <div class="event-priority priority-${priorityClass}">${priorityClass.toUpperCase()}</div>
        </div>
        <div class="event-title">${event.title}</div>
        ${event.description ? `<div class="event-description">${event.description}</div>` : ''}
        <div class="event-meta">
          <div class="event-category" style="background-color: ${categoryColor}">${categoryName}</div>
          ${metaInfo ? `<div class="event-meta-info">${metaInfo}</div>` : ''}
        </div>
        ${tagsHtml}
      </div>
    `;
  }

  private getAllEvents(): (CalendarEvent & { calendarId: string })[] {
    const stored = localStorage.getItem(this.eventsKey);
    return stored ? JSON.parse(stored) : [];
  }

  public exportCalendar(id: string): string {
    const calendar = this.getCalendar(id);
    if (!calendar) {
      throw new Error('Calendar not found');
    }

    const events = this.getEvents(id);
    const exportData = {
      calendar,
      events,
      categories: this.categoryManager.getCategories(),
      tags: this.categoryManager.getTags(),
      exportDate: new Date().toISOString(),
    };

    return JSON.stringify(exportData, null, 2);
  }

  public importCalendar(data: string): Calendar {
    try {
      const importData = JSON.parse(data);
      const calendar = importData.calendar;
      const events = importData.events || [];
      const categories = importData.categories || [];
      const tags = importData.tags || [];

      // Импортируем категории и теги
      categories.forEach((cat: any) => {
        this.categoryManager.createCategory(cat.name, cat.color);
      });

      tags.forEach((tag: any) => {
        this.categoryManager.createTag(tag.name, tag.color);
      });

      const newCalendar = this.createCalendar({
        title: calendar.title,
        description: calendar.description,
      });

      events.forEach((eventData: any) => {
        this.createEvent(eventData, newCalendar.id);
      });

      return newCalendar;
    } catch (error) {
      throw new Error('Invalid calendar data format');
    }
  }

  public copyCalendar(id: string): Calendar {
    const originalCalendar = this.getCalendar(id);
    if (!originalCalendar) {
      throw new Error('Calendar not found');
    }

    const events = this.getEvents(id);
    const newCalendar = this.createCalendar({
      title: `${originalCalendar.title} (Copy)`,
      description: originalCalendar.description,
    });

    // Копируем события
    events.forEach((event) => {
      this.createEvent(
        {
          title: event.title,
          description: event.description,
          date: event.date,
          time: event.time,
          duration: event.duration,
          location: event.location,
          color: event.color,
          isAllDay: event.isAllDay,
          priority: event.priority,
          category: event.category,
          tags: event.tags,
          attendees: event.attendees,
          reminder: event.reminder,
          recurring: event.recurring,
          attachments: event.attachments,
        },
        newCalendar.id
      );
    });

    return newCalendar;
  }

  public copyEvent(id: string): CalendarEvent {
    const originalEvent = this.getEvent(id);
    if (!originalEvent) {
      throw new Error('Event not found');
    }

    const allEvents = this.getAllEvents();
    const eventWithCalendar = allEvents.find((e) => e.id === id);
    if (!eventWithCalendar) {
      throw new Error('Event calendar not found');
    }

    return this.createEvent(
      {
        title: `${originalEvent.title} (Copy)`,
        description: originalEvent.description,
        date: originalEvent.date,
        time: originalEvent.time,
        duration: originalEvent.duration,
        location: originalEvent.location,
        color: originalEvent.color,
        isAllDay: originalEvent.isAllDay,
        priority: originalEvent.priority,
        category: originalEvent.category,
        tags: originalEvent.tags,
        attendees: originalEvent.attendees,
        reminder: originalEvent.reminder,
        recurring: originalEvent.recurring,
        attachments: originalEvent.attachments,
      },
      eventWithCalendar.calendarId
    );
  }
}
