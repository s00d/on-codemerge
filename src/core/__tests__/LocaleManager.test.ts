import { LocaleManager } from '../services/LocaleManager';

describe('LocaleManager', () => {
  let localeManager: LocaleManager;

  beforeEach(() => {
    localeManager = new LocaleManager('en');
  });

  it('should load the default locale', async () => {
    await localeManager.loadLocale('en');
    expect(localeManager.getCurrentLocale()).toMatchSnapshot();
  });

  it('should translate a key correctly', async () => {
    await localeManager.loadLocale('en');
    const translation = localeManager.translate('New Block');
    expect(translation).toMatchSnapshot();
  });

  it('should fallback to default locale if translation is missing', async () => {
    await localeManager.loadLocale('ru');
    const translation = localeManager.translate('NonExistentKey');
    expect(translation).toMatchSnapshot();
  });

  it('should replace placeholders in translations', async () => {
    await localeManager.loadLocale('en');
    const translation = localeManager.translate('File size exceeds {{max}} limit', { max: '10MB' });
    expect(translation).toMatchSnapshot();
  });

  it('should set a new locale', async () => {
    await localeManager.loadLocale('ru');
    await localeManager.setLocale('ru');
    expect(localeManager.getCurrentLocale()).toMatchSnapshot();
  });

  it('should return the list of loaded locales', async () => {
    await localeManager.loadLocale('en');
    await localeManager.loadLocale('ru');
    const loadedLocales = localeManager.getLoadedLocales();
    expect(loadedLocales).toMatchSnapshot();
  });
});
