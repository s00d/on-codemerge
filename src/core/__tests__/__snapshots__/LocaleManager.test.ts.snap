// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleManager should fallback to default locale if translation is missing 1`] = `"NonExistentKey"`;

exports[`LocaleManager should load the default locale 1`] = `"en"`;

exports[`LocaleManager should replace placeholders in translations 1`] = `"File size exceeds 10MB limit"`;

exports[`LocaleManager should return the list of loaded locales 1`] = `
[
  "en",
  "ru",
]
`;

exports[`LocaleManager should set a new locale 1`] = `"en"`;

exports[`LocaleManager should translate a key correctly 1`] = `"New Block"`;
